lib/internal/streams/async_iterator.js

```diff
@@ -0,0 +1,132 @@
+'use strict';
+
+const kLastResolve = Symbol('lastResolve');
+const kLastReject = Symbol('lastReject');
+const kError = Symbol('error');
+const kEnded = Symbol('ended');
+const kLastPromise = Symbol('lastPromise');
+const kHandlePromise = Symbol('handlePromise');
+
+class AsyncIteratorRecord {
+  constructor(value, done) {
+    this.done = done;
+    this.value = value;
+  }
+}
+
+function readAndResolve(iter) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    const data = iter.stream.read();
+    // we defer if data is null
+    // we can be expecting either 'end' or
+    // 'error'
+    if (data !== null) {
+      resolve(new AsyncIteratorRecord(data, false));
+      iter[kLastResolve] = null;
+      iter[kLastReject] = null;
+    }
+  }
+}
+
+function onReadable(iter) {
+  // we wait for the next tick, because it might
+  // emit an error with process.nextTick
+  process.nextTick(readAndResolve, iter);
+}
+
+function onEnd(iter) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    resolve(new AsyncIteratorRecord(null, true));
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter[kEnded] = true;
+}
+
+function onError(iter, err) {
+  const reject = iter[kLastReject];
+  // reject if we are waiting for data in the Promise
+  // returned by next() and store the error
+  if (reject !== null) {
+    reject(err);
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter.error = err;
+}
+
+class ReadableAsyncIterator {
+  constructor(stream) {
+    this.stream = stream;
```

Comment:
We might want to not expose this property either (like the other ones).

Reply:
I left this exposed on purpose. I think it might be useful if the AsyncIterator is used directly. I'll add a comment.

Reply:
What if we make it a getter and store the stream internally? Because setting this will not have the effect someone might think.